src/rust-crypto/secret-storage.ts

```diff
@@ -0,0 +1,42 @@
+/*
+Copyright 2023 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { ServerSideSecretStorage } from "../secret-storage";
+
+/**
+ * Check that the private cross signing keys (master, self signing, user signing) are stored into the secret storage and encrypted with the secret storage key.
+ *
+ * @param secretStorage - The secret store using account data
+ * @returns True if one of the secret storage master keys is shared with the secret storage user signing and self signing keys.
+ */
+export async function secretStorageContainsCrossSigningKeys(secretStorage: ServerSideSecretStorage): Promise<boolean> {
+    // Get the secret storage keys stored into the secret storage
+    const secretStorageMasterKeys = await secretStorage.isStored("m.cross_signing.master");
+
+    // Not stored keys
+    if (!secretStorageMasterKeys) return false;
+
+    // Get the user signing keys stored into the secret storage
+    const secretStorageUserSigningKeys = (await secretStorage.isStored(`m.cross_signing.user_signing`)) || {};
+    // Get the self signing keys stored into the secret storage
+    const secretStorageSelfSigningKeys = (await secretStorage.isStored(`m.cross_signing.self_signing`)) || {};
+
+    // Check that one of the secret storage master keys is shared with the secret storage user signing and self signing keys
```

Comment:
I think you're confusing cross-signing keys and secret-storage keys again.

To be clear:
 * We have a [master cross-signing key](https://spec.matrix.org/v1.7/client-server-api/#cross-signing). It is an ed25519 key, which means it is asymmetric (it is a key *pair*, with a public part and a private part), and it can be used for *signing* things but not encrypting them. It is used to sign the user-signing and self-signing cross-signing keys.
 * We do *not* have a master secret-storage key (or at least, we don't use that term). Instead, we just have [secret storage keys](https://spec.matrix.org/v1.7/client-server-api/#storage). There is some [fancy maths](https://spec.matrix.org/v1.7/client-server-api/#msecret_storagev1aes-hmac-sha2) but basically they are AES keys - so it is a *symmetric* key (there is no "public" part), and is used for *encryption*, not signing. They can nominally be used for encrypting any secret, but in this case we are using them to encrypt the private parts of the cross-signing keys.


```suggestion
    // Check that one of the secret storage keys used to encrypt the master key was also used to encrypt the user-signing and self-signing keys
```